write-fonts/src/graph/splitting/mark2base.rs

```diff
@@ -0,0 +1,680 @@
+        .iter()
+        .enumerate()
+        .filter_map(|(i, gid)| mark_glyphs_by_cov_id.contains(&i).then_some(gid))
+        .collect();
+    let new_mark_coverage = super::make_table_data(&new_mark_coverage);
+    let new_mark_array = split_off_mark_array(old_mark_array, start as u16, &mark_glyphs_by_cov_id);
+    let new_base_array = split_off_base_array(old_base_array, start, end, class_info.len());
+    let new_mark_coverage_id = graph.add_object(new_mark_coverage);
+    let new_mark_array_id = graph.add_object(new_mark_array);
+    let new_base_array_id = graph.add_object(new_base_array);
+
+    new_subtable.write(1u16); // format
+    new_subtable.add_offset(new_mark_coverage_id, 2, 0);
+    new_subtable.add_offset(base_coverage_id, 2, 0);
+    new_subtable.write(mark_class_count);
+    new_subtable.add_offset(new_mark_array_id, 2, 0);
+    new_subtable.add_offset(new_base_array_id, 2, 0);
+
+    new_subtable
+}
+
+// <https://github.com/harfbuzz/harfbuzz/blob/f26fd69d858642d76413b8f/src/graph/markbasepos-graph.hh#L170>
+fn split_off_mark_array(
+    old_mark_array: &TableData,
+    first_class: u16,
+    mark_glyph_coverage_ids: &HashSet<usize>,
+) -> TableData {
+    let mark_array = old_mark_array.reparse::<rgpos::MarkArray>().unwrap();
+    let mark_count = mark_glyph_coverage_ids.len() as u16;
+
+    let mut new_mark_array = TableData::new(old_mark_array.type_);
+    new_mark_array.write(mark_count);
+
+    for (i, mark_record) in mark_array.mark_records().iter().enumerate() {
+        if !mark_glyph_coverage_ids.contains(&i) {
+            continue;
+        }
+        let new_class = mark_record.mark_class() - first_class;
+        let anchor_offset = old_mark_array.offsets[i].object;
+
+        new_mark_array.write(new_class);
+        new_mark_array.add_offset(anchor_offset, 2, 0);
+    }
+
+    new_mark_array
+}
+
+fn split_off_base_array(
+    old_base_array: &TableData,
+    start: usize,
+    end: usize,
+    old_mark_class_count: usize,
+) -> TableData {
+    let mut new_base_array = TableData::new(old_base_array.type_);
+    let base_count: u16 = old_base_array.read_at(0).unwrap_or(0);
+    new_base_array.write(base_count);
+
+    // the base array contains a (base_count x mark_count) matrix of offsets.
+    // for each base, we want to prune the marks to only include those
+    // in the range (start..end).
+
+    let base_array: rgpos::BaseArray = old_base_array
+        .reparse_with_args(&(old_mark_class_count as u16))
+        .unwrap();
+
+    let mut next_offset_idx = 0;
+    // because offsets may be null, and there is no pattern, we visit each one
+    for base_record in base_array.base_records().iter() {
+        let base_record = base_record.unwrap();
+        for (mark_class, offset) in base_record.base_anchor_offsets().iter().enumerate() {
+            let has_offset = !offset.get().is_null();
+            let in_range = (start..end).contains(&mark_class);
```

Comment:
I think this comment should be on the inner for loop. 

I'm assuming this means that we iterate all and check the range rather than slicing `base_anchor_offsets()` on `start..end` because we need to track `next_offset_idx`? Could we break out of the loop when `mark_class >= end`? (not suggesting we do so, just checking my understanding of what's going on)

Reply:
noted, reading this back I agree that there could be more explanation here. Basically: what is tricky here is that the anchor tables may or may not have subtables, and there is no pattern around how subtables are distributed across the set of anchor tables, so we need to visit each offset and see if it's null or not to figure out whether we are copying over a subtable.

The reason we can't break early is because even if we're out of range, if we saw an offset we need to increment `next_offset_idx` so that once we're back in range we are copying the correct subtable. 🤕

Reply:
Ohh, yep, I see it now. The comment seems like it's in the proper place then... I was just missing that the offset list is associated with the base array and not the base record (obvious in hindsigh). Thanks!